{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Module 13-2 Learning Notebook: Column transformation and Pipelines\n",
    "\n",
    "This gets a little more complicated, but if you can master both the ColumnTransformer & the Pipeline, you can write very short, efficient code to explore many algorithms quickly.\n",
    "\n",
    "**Data:**\n",
    "    \n",
    "We'll use the same 'gene expression' dataset, except I've included a 'gender' column with it.\n",
    "\n",
    "The data used in this problem is a simplified version of using \"gene expression\" to predict cancer in people. It is \n",
    "based on this dataset:\n",
    "\n",
    "    http://archive.ics.uci.edu/ml/datasets/gene+expression+cancer+RNA-Seq\n",
    "    \n",
    "**Method:**\n",
    "1. Load the data\n",
    "2. Introduction to a column transformer\n",
    "3. Column Transformer missing values (impute)\n",
    "4. Convert categories to numbers\n",
    "5. Putting is all togehter: do all transfomers at once\n",
    "6. Combine ColumnTransformer with the use of the Pipeline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.preprocessing import StandardScaler, MinMaxScaler\n",
    "from sklearn.preprocessing import LabelEncoder, OrdinalEncoder\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.compose import make_column_transformer,ColumnTransformer\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.pipeline import Pipeline\n",
    "import boto3\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# Prevent pandas from displaying in scientific notation\n",
    "pd.set_option('display.float_format', lambda x: '%.3f' % x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Load the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "      <th>gender</th>\n",
       "      <th>cancer_detected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.759</td>\n",
       "      <td>27.342</td>\n",
       "      <td>118.878</td>\n",
       "      <td>-29.800</td>\n",
       "      <td>641.214</td>\n",
       "      <td>-12.906</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.727</td>\n",
       "      <td>16.191</td>\n",
       "      <td>122.520</td>\n",
       "      <td>-56.616</td>\n",
       "      <td>239.289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.235</td>\n",
       "      <td>19.346</td>\n",
       "      <td>128.828</td>\n",
       "      <td>-90.479</td>\n",
       "      <td>374.460</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.922</td>\n",
       "      <td>20.417</td>\n",
       "      <td>57.907</td>\n",
       "      <td>-62.898</td>\n",
       "      <td>398.819</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.228</td>\n",
       "      <td>26.416</td>\n",
       "      <td>87.028</td>\n",
       "      <td>-38.963</td>\n",
       "      <td>581.078</td>\n",
       "      <td>26.624</td>\n",
       "      <td>female</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene1  gene2   gene3   gene4   gene5   gene6  gender  cancer_detected\n",
       "0  0.759 27.342 118.878 -29.800 641.214 -12.906    male                0\n",
       "1  3.727 16.191 122.520 -56.616 239.289     NaN    male                1\n",
       "2  2.235 19.346 128.828 -90.479 374.460     NaN  female                1\n",
       "3  4.922 20.417  57.907 -62.898 398.819     NaN    male                0\n",
       "4  1.228 26.416  87.028 -38.963 581.078  26.624  female                1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load df from S3 .csv\n",
    "sess = boto3.session.Session()\n",
    "s3 = sess.client('s3') \n",
    "source_bucket = 'machinelearning-read-only'\n",
    "source_key = 'data/gene-cancer-gender.csv'\n",
    "response = s3.get_object(Bucket=source_bucket, Key=source_key)\n",
    "df = pd.read_csv(response.get(\"Body\"))\n",
    "# Notice scales, missing data and categorical (string) data\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "      <th>cancer_detected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>75.000</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>4.118</td>\n",
       "      <td>20.732</td>\n",
       "      <td>82.765</td>\n",
       "      <td>-50.003</td>\n",
       "      <td>408.900</td>\n",
       "      <td>114.232</td>\n",
       "      <td>0.320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.540</td>\n",
       "      <td>4.111</td>\n",
       "      <td>32.476</td>\n",
       "      <td>21.673</td>\n",
       "      <td>164.819</td>\n",
       "      <td>47.338</td>\n",
       "      <td>0.469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-0.253</td>\n",
       "      <td>10.724</td>\n",
       "      <td>15.323</td>\n",
       "      <td>-127.540</td>\n",
       "      <td>26.891</td>\n",
       "      <td>-12.906</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.238</td>\n",
       "      <td>17.777</td>\n",
       "      <td>61.068</td>\n",
       "      <td>-60.887</td>\n",
       "      <td>284.872</td>\n",
       "      <td>86.485</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>4.588</td>\n",
       "      <td>20.462</td>\n",
       "      <td>81.360</td>\n",
       "      <td>-48.005</td>\n",
       "      <td>407.237</td>\n",
       "      <td>120.822</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>5.174</td>\n",
       "      <td>23.643</td>\n",
       "      <td>106.616</td>\n",
       "      <td>-39.195</td>\n",
       "      <td>526.622</td>\n",
       "      <td>152.787</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>7.147</td>\n",
       "      <td>29.389</td>\n",
       "      <td>150.491</td>\n",
       "      <td>9.134</td>\n",
       "      <td>813.139</td>\n",
       "      <td>206.951</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene1   gene2   gene3    gene4   gene5   gene6  cancer_detected\n",
       "count 100.000 100.000 100.000  100.000 100.000  75.000          100.000\n",
       "mean    4.118  20.732  82.765  -50.003 408.900 114.232            0.320\n",
       "std     1.540   4.111  32.476   21.673 164.819  47.338            0.469\n",
       "min    -0.253  10.724  15.323 -127.540  26.891 -12.906            0.000\n",
       "25%     3.238  17.777  61.068  -60.887 284.872  86.485            0.000\n",
       "50%     4.588  20.462  81.360  -48.005 407.237 120.822            0.000\n",
       "75%     5.174  23.643 106.616  -39.195 526.622 152.787            1.000\n",
       "max     7.147  29.389 150.491    9.134 813.139 206.951            1.000"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe() # This won't include 'gender' since it is not numeric\n",
    "# Notice gene6 is missing values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Introduction to a column transformer\n",
    "\n",
    "Applies transformers to one or more columns of pandas DataFrame.\n",
    "\n",
    "This estimator allows different columns or column subsets of the input to be transformed separately and the features generated by each transformer will be concatenated to form a single feature space. This is useful for heterogeneous or columnar data, to combine several feature extraction mechanisms or transformations into a single transformer."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Review: scale the whole dataset using a scaler:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>75.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.591</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.499</td>\n",
       "      <td>0.567</td>\n",
       "      <td>0.486</td>\n",
       "      <td>0.578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.208</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.159</td>\n",
       "      <td>0.210</td>\n",
       "      <td>0.215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.472</td>\n",
       "      <td>0.378</td>\n",
       "      <td>0.338</td>\n",
       "      <td>0.488</td>\n",
       "      <td>0.328</td>\n",
       "      <td>0.452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.654</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.489</td>\n",
       "      <td>0.582</td>\n",
       "      <td>0.484</td>\n",
       "      <td>0.608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.733</td>\n",
       "      <td>0.692</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.646</td>\n",
       "      <td>0.636</td>\n",
       "      <td>0.754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene1   gene2   gene3   gene4   gene5  gene6\n",
       "count 100.000 100.000 100.000 100.000 100.000 75.000\n",
       "mean    0.591   0.536   0.499   0.567   0.486  0.578\n",
       "std     0.208   0.220   0.240   0.159   0.210  0.215\n",
       "min     0.000   0.000   0.000   0.000   0.000  0.000\n",
       "25%     0.472   0.378   0.338   0.488   0.328  0.452\n",
       "50%     0.654   0.522   0.489   0.582   0.484  0.608\n",
       "75%     0.733   0.692   0.675   0.646   0.636  0.754\n",
       "max     1.000   1.000   1.000   1.000   1.000  1.000"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Recall from previous lesson on scaling data:\n",
    "#\n",
    "X = df.drop(['gender', 'cancer_detected'], axis = 1) # Drop some columns\n",
    "# This creates a scaler\n",
    "norm_scaler = MinMaxScaler()\n",
    "# Compute the minimum and maximum to be used for later scaling.\n",
    "norm_scaler.fit(X)\n",
    "# Do the scaling, this returns a numpy array\n",
    "norm_scaled_array = norm_scaler.transform(X) \n",
    "# Create a new data frame from the scaled values\n",
    "X_scaled = pd.DataFrame(data = norm_scaled_array, columns = X.columns)\n",
    "X_scaled.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Scale only one or a set of columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.759</td>\n",
       "      <td>27.342</td>\n",
       "      <td>118.878</td>\n",
       "      <td>-29.800</td>\n",
       "      <td>641.214</td>\n",
       "      <td>-12.906</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.727</td>\n",
       "      <td>16.191</td>\n",
       "      <td>122.520</td>\n",
       "      <td>-56.616</td>\n",
       "      <td>239.289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.235</td>\n",
       "      <td>19.346</td>\n",
       "      <td>128.828</td>\n",
       "      <td>-90.479</td>\n",
       "      <td>374.460</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene1  gene2   gene3   gene4   gene5   gene6  gender\n",
       "0  0.759 27.342 118.878 -29.800 641.214 -12.906    male\n",
       "1  3.727 16.191 122.520 -56.616 239.289     NaN    male\n",
       "2  2.235 19.346 128.828 -90.479 374.460     NaN  female"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We can do it only on a column or a set of columns\n",
    "#\n",
    "# Create a fresh set of the data, only withhold the target column\n",
    "X = df.drop('cancer_detected', axis = 1)\n",
    "X.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ColumnTransformer(n_jobs=None, remainder='passthrough', sparse_threshold=0.3,\n",
       "                  transformer_weights=None,\n",
       "                  transformers=[('norm',\n",
       "                                 MinMaxScaler(copy=True, feature_range=(0, 1)),\n",
       "                                 ['gene1', 'gene2', 'gene3']),\n",
       "                                ('stand',\n",
       "                                 StandardScaler(copy=True, with_mean=True,\n",
       "                                                with_std=True),\n",
       "                                 ['gene4', 'gene5'])],\n",
       "                  verbose=False)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.compose import ColumnTransformer\n",
    "\n",
    "# Identify columns to scale in different ways\n",
    "norm_columns = ['gene1', 'gene2', 'gene3']\n",
    "stand_columns = ['gene4', 'gene5'] # Leave gene6 alone for now\n",
    "\n",
    "# Create the column transformer using the right scaler\n",
    "ct = ColumnTransformer(\n",
    "    transformers = [\n",
    "    ('norm', MinMaxScaler(), norm_columns), # 0-1\n",
    "    ('stand', StandardScaler(), stand_columns) # Mean = 0, std_dev = 1\n",
    "    ], \n",
    "    remainder = 'passthrough') # Let the other columns ('gender') just passthrough\n",
    "\n",
    "# Prepare the ct to transform the data\n",
    "ct.fit(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.13682382844443777 0.890330448269027 0.7661204972831208\n",
      " 0.9368696470604763 1.416612146906083 -12.90552460113543 'male'] \n",
      "\n",
      "[0.5378644677950846 0.2928779251687599 0.7930609165447469\n",
      " -0.30666730808115206 -1.0342599583793624 nan 'male'] \n",
      "\n",
      "[0.33618422107090584 0.4619156392077014 0.8397265238705481\n",
      " -1.877005328467732 -0.21001375265276553 nan 'female'] \n",
      "\n",
      "[0.6994323337892093 0.5192903128294121 0.31503948534484494\n",
      " -0.5979689019582491 -0.0614749160394467 nan 'male'] \n",
      "\n",
      "[0.20015209591008581 0.8407037029387529 0.5304836118863425\n",
      " 0.5119879279002758 1.0499116216970483 26.6243243517391 'female'] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Perform the transformation\n",
    "transformed_array = ct.transform(X)  # Returns a numpy array\n",
    "# Print the first 5\n",
    "for row in transformed_array[0:5]:\n",
    "    print(row,'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>75.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.591</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.499</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>114.232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.208</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.240</td>\n",
       "      <td>1.005</td>\n",
       "      <td>1.005</td>\n",
       "      <td>47.338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-3.596</td>\n",
       "      <td>-2.329</td>\n",
       "      <td>-12.906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.472</td>\n",
       "      <td>0.378</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>-0.756</td>\n",
       "      <td>86.485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.654</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.489</td>\n",
       "      <td>0.093</td>\n",
       "      <td>-0.010</td>\n",
       "      <td>120.822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.733</td>\n",
       "      <td>0.692</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.501</td>\n",
       "      <td>0.718</td>\n",
       "      <td>152.787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2.742</td>\n",
       "      <td>2.465</td>\n",
       "      <td>206.951</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene1   gene2   gene3   gene4   gene5   gene6\n",
       "count 100.000 100.000 100.000 100.000 100.000  75.000\n",
       "mean    0.591   0.536   0.499   0.000  -0.000 114.232\n",
       "std     0.208   0.220   0.240   1.005   1.005  47.338\n",
       "min     0.000   0.000   0.000  -3.596  -2.329 -12.906\n",
       "25%     0.472   0.378   0.338  -0.505  -0.756  86.485\n",
       "50%     0.654   0.522   0.489   0.093  -0.010 120.822\n",
       "75%     0.733   0.692   0.675   0.501   0.718 152.787\n",
       "max     1.000   1.000   1.000   2.742   2.465 206.951"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is typically not necessary, but just to show the results, create a dataframe from the array\n",
    "X_transformed_df = pd.DataFrame(data = transformed_array, columns = X.columns)\n",
    "# Need to convert these columns from objects to floats:\n",
    "X_transformed_df = X_transformed_df.apply(pd.to_numeric, errors='ignore')\n",
    "X_transformed_df.describe() # Ignore 'gender' for now. It is not a numeric column\n",
    "# We have successfuly scaled columns 1-3 (normalized) and 4-5 (standardized), 'gene6' and 'gender' was ignored"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Column Transformer missing values (impute)\n",
    "We are missing 25 rows in the 'gene6' column<P>\n",
    "    \n",
    "Be aware of how the transfomrd array is returned below. It is out of order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[-12.90552460113543 0.7593336360642229 27.342287359757897\n",
      " 118.87838370542934 -29.800470404715927 641.2144908190065 'male'] \n",
      "\n",
      "[114.23174731226524 3.7269018931959352 16.190668631240115\n",
      " 122.51987028373456 -56.61609162376188 239.2890296354452 'male'] \n",
      "\n",
      "[114.23174731226524 2.23453468762236 19.34580492210508 128.82757400189527\n",
      " -90.47884846942158 374.45950044501 'female'] \n",
      "\n",
      "[114.23174731226524 4.92245073152067 20.41671927958799 57.90659871761008\n",
      " -62.897716926182376 398.8188051515788 'male'] \n",
      "\n",
      "[26.6243243517391 1.2279418963505029 26.415990254527344 87.02778230264246\n",
      " -38.96261648874555 581.0782326303199 'female'] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Impute the missing values using the column transformer\n",
    "impute_columns = ['gene6']\n",
    "\n",
    "# Create the column transformer using the right scaler\n",
    "impute_ct = ColumnTransformer(    \n",
    "    transformers = \n",
    "    [\n",
    "    # This is the same imputer we used back in the lesson on imputing\n",
    "    ('impute', SimpleImputer(missing_values=np.nan, strategy='mean'), impute_columns)\n",
    "    ],\n",
    "    remainder = 'passthrough',\n",
    "    )\n",
    "\n",
    "# Prepare the ct to transform the data\n",
    "impute_ct.fit(X)\n",
    "# Perform the transformation\n",
    "transformed_array = impute_ct.transform(X)  # Returns a numpy array\n",
    "# Show the first 5 rows\n",
    "for row in transformed_array[0:5]:\n",
    "    print(row,'\\n')\n",
    "# This numpy array is all the columns, but they are out of order:\n",
    "#    They are ordered like this: gene6, gene1, gene2, gene3, gene4, gene5, gender\n",
    "#    This is because the transformed column is returned first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene6</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-12.906</td>\n",
       "      <td>0.759</td>\n",
       "      <td>27.342</td>\n",
       "      <td>118.878</td>\n",
       "      <td>-29.800</td>\n",
       "      <td>641.214</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>114.232</td>\n",
       "      <td>3.727</td>\n",
       "      <td>16.191</td>\n",
       "      <td>122.520</td>\n",
       "      <td>-56.616</td>\n",
       "      <td>239.289</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>114.232</td>\n",
       "      <td>2.235</td>\n",
       "      <td>19.346</td>\n",
       "      <td>128.828</td>\n",
       "      <td>-90.479</td>\n",
       "      <td>374.460</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>114.232</td>\n",
       "      <td>4.922</td>\n",
       "      <td>20.417</td>\n",
       "      <td>57.907</td>\n",
       "      <td>-62.898</td>\n",
       "      <td>398.819</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>26.624</td>\n",
       "      <td>1.228</td>\n",
       "      <td>26.416</td>\n",
       "      <td>87.028</td>\n",
       "      <td>-38.963</td>\n",
       "      <td>581.078</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene6 gene1  gene2   gene3   gene4   gene5  gender\n",
       "0 -12.906 0.759 27.342 118.878 -29.800 641.214    male\n",
       "1 114.232 3.727 16.191 122.520 -56.616 239.289    male\n",
       "2 114.232 2.235 19.346 128.828 -90.479 374.460  female\n",
       "3 114.232 4.922 20.417  57.907 -62.898 398.819    male\n",
       "4  26.624 1.228 26.416  87.028 -38.963 581.078  female"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Just to show the results, create a dataframe from the array\n",
    "new_cols = ['gene6', 'gene1','gene2','gene3','gene4','gene5','gender']\n",
    "X_imputed_df = pd.DataFrame(data = transformed_array, columns = new_cols)\n",
    "X_imputed_df.head(5) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene6</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>76.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>114.232</td>\n",
       "      <td>0.759</td>\n",
       "      <td>27.342</td>\n",
       "      <td>118.878</td>\n",
       "      <td>-29.800</td>\n",
       "      <td>641.214</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>25.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene6   gene1   gene2   gene3   gene4   gene5  gender\n",
       "count  100.000 100.000 100.000 100.000 100.000 100.000     100\n",
       "unique  76.000 100.000 100.000 100.000 100.000 100.000       2\n",
       "top    114.232   0.759  27.342 118.878 -29.800 641.214  female\n",
       "freq    25.000   1.000   1.000   1.000   1.000   1.000      55"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Why only 4 stats here?\n",
    "X_imputed_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene6     object\n",
       "gene1     object\n",
       "gene2     object\n",
       "gene3     object\n",
       "gene4     object\n",
       "gene5     object\n",
       "gender    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here is why: \n",
    "X_imputed_df.dtypes # The data types got change to objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene6</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>114.232</td>\n",
       "      <td>4.118</td>\n",
       "      <td>20.732</td>\n",
       "      <td>82.765</td>\n",
       "      <td>-50.003</td>\n",
       "      <td>408.900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>40.927</td>\n",
       "      <td>1.540</td>\n",
       "      <td>4.111</td>\n",
       "      <td>32.476</td>\n",
       "      <td>21.673</td>\n",
       "      <td>164.819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-12.906</td>\n",
       "      <td>-0.253</td>\n",
       "      <td>10.724</td>\n",
       "      <td>15.323</td>\n",
       "      <td>-127.540</td>\n",
       "      <td>26.891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>99.475</td>\n",
       "      <td>3.238</td>\n",
       "      <td>17.777</td>\n",
       "      <td>61.068</td>\n",
       "      <td>-60.887</td>\n",
       "      <td>284.872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>114.232</td>\n",
       "      <td>4.588</td>\n",
       "      <td>20.462</td>\n",
       "      <td>81.360</td>\n",
       "      <td>-48.005</td>\n",
       "      <td>407.237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>143.868</td>\n",
       "      <td>5.174</td>\n",
       "      <td>23.643</td>\n",
       "      <td>106.616</td>\n",
       "      <td>-39.195</td>\n",
       "      <td>526.622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>206.951</td>\n",
       "      <td>7.147</td>\n",
       "      <td>29.389</td>\n",
       "      <td>150.491</td>\n",
       "      <td>9.134</td>\n",
       "      <td>813.139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene6   gene1   gene2   gene3    gene4   gene5\n",
       "count 100.000 100.000 100.000 100.000  100.000 100.000\n",
       "mean  114.232   4.118  20.732  82.765  -50.003 408.900\n",
       "std    40.927   1.540   4.111  32.476   21.673 164.819\n",
       "min   -12.906  -0.253  10.724  15.323 -127.540  26.891\n",
       "25%    99.475   3.238  17.777  61.068  -60.887 284.872\n",
       "50%   114.232   4.588  20.462  81.360  -48.005 407.237\n",
       "75%   143.868   5.174  23.643 106.616  -39.195 526.622\n",
       "max   206.951   7.147  29.389 150.491    9.134 813.139"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert them back to floats so we can look at stats\n",
    "# Remember, our goal was to impute missing values for 'gene6'\n",
    "X_imputed_df = X_imputed_df.apply(pd.to_numeric, errors='ignore') # This dropped 'gender'\n",
    "X_imputed_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We have successfully imputed missing values in 'gene6'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Convert categories to numbers\n",
    "Convert a category 'string' column to a numeric column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.759</td>\n",
       "      <td>27.342</td>\n",
       "      <td>118.878</td>\n",
       "      <td>-29.800</td>\n",
       "      <td>641.214</td>\n",
       "      <td>-12.906</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.727</td>\n",
       "      <td>16.191</td>\n",
       "      <td>122.520</td>\n",
       "      <td>-56.616</td>\n",
       "      <td>239.289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.235</td>\n",
       "      <td>19.346</td>\n",
       "      <td>128.828</td>\n",
       "      <td>-90.479</td>\n",
       "      <td>374.460</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.922</td>\n",
       "      <td>20.417</td>\n",
       "      <td>57.907</td>\n",
       "      <td>-62.898</td>\n",
       "      <td>398.819</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.228</td>\n",
       "      <td>26.416</td>\n",
       "      <td>87.028</td>\n",
       "      <td>-38.963</td>\n",
       "      <td>581.078</td>\n",
       "      <td>26.624</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene1  gene2   gene3   gene4   gene5   gene6  gender\n",
       "0  0.759 27.342 118.878 -29.800 641.214 -12.906    male\n",
       "1  3.727 16.191 122.520 -56.616 239.289     NaN    male\n",
       "2  2.235 19.346 128.828 -90.479 374.460     NaN  female\n",
       "3  4.922 20.417  57.907 -62.898 398.819     NaN    male\n",
       "4  1.228 26.416  87.028 -38.963 581.078  26.624  female"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get a fresh copy of the data\n",
    "X = df.drop('cancer_detected', axis = 1)\n",
    "X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[  0.           0.75933364  27.34228736 118.87838371 -29.8004704\n",
      " 641.21449082 -12.9055246 ] \n",
      "\n",
      "[  0.           3.72690189  16.19066863 122.51987028 -56.61609162\n",
      " 239.28902964          nan] \n",
      "\n",
      "[  1.           2.23453469  19.34580492 128.827574   -90.47884847\n",
      " 374.45950045          nan] \n",
      "\n",
      "[  0.           4.92245073  20.41671928  57.90659872 -62.89771693\n",
      " 398.81880515          nan] \n",
      "\n",
      "[  1.           1.2279419   26.41599025  87.0277823  -38.96261649\n",
      " 581.07823263  26.62432435] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "category_columns = ['gender']\n",
    "#\n",
    "# Create the column transformer using the OrdinalEncoder\n",
    "#   https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html\n",
    "#\n",
    "category_ct = ColumnTransformer(\n",
    "    transformers = [\n",
    "    ('cat', OrdinalEncoder(categories = [['male','female']]), category_columns)\n",
    "    ],\n",
    "    remainder = 'passthrough')\n",
    "\n",
    "# Prepare the ct to transform the data\n",
    "category_ct.fit(X)\n",
    "# perform the transformation\n",
    "transformed_array = category_ct.transform(X)\n",
    "for row in transformed_array[0:5]:\n",
    "    print(row,'\\n')\n",
    "# Gender is now in the first column, since it was transformed, the other columns were passed though."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.759</td>\n",
       "      <td>27.342</td>\n",
       "      <td>118.878</td>\n",
       "      <td>-29.800</td>\n",
       "      <td>641.214</td>\n",
       "      <td>-12.906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>3.727</td>\n",
       "      <td>16.191</td>\n",
       "      <td>122.520</td>\n",
       "      <td>-56.616</td>\n",
       "      <td>239.289</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.000</td>\n",
       "      <td>2.235</td>\n",
       "      <td>19.346</td>\n",
       "      <td>128.828</td>\n",
       "      <td>-90.479</td>\n",
       "      <td>374.460</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000</td>\n",
       "      <td>4.922</td>\n",
       "      <td>20.417</td>\n",
       "      <td>57.907</td>\n",
       "      <td>-62.898</td>\n",
       "      <td>398.819</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.228</td>\n",
       "      <td>26.416</td>\n",
       "      <td>87.028</td>\n",
       "      <td>-38.963</td>\n",
       "      <td>581.078</td>\n",
       "      <td>26.624</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gender  gene1  gene2   gene3   gene4   gene5   gene6\n",
       "0   0.000  0.759 27.342 118.878 -29.800 641.214 -12.906\n",
       "1   0.000  3.727 16.191 122.520 -56.616 239.289     NaN\n",
       "2   1.000  2.235 19.346 128.828 -90.479 374.460     NaN\n",
       "3   0.000  4.922 20.417  57.907 -62.898 398.819     NaN\n",
       "4   1.000  1.228 26.416  87.028 -38.963 581.078  26.624"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a dataframe from the array\n",
    "new_cols = ['gender','gene1','gene2','gene3','gene4','gene5','gene6']\n",
    "X_cat_df = pd.DataFrame(data = transformed_array, columns = new_cols)\n",
    "X_cat_df.head()\n",
    "# Successfully transformed gender to 0 = 'male' and 1 = 'female'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Putting is all togehter: do all transfomers at once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fresh dataframe\n",
    "X = df.drop('cancer_detected', axis = 1)\n",
    "#\n",
    "# First, the normalized columns\n",
    "norm_columns = ['gene1', 'gene2', 'gene3']\n",
    "norm_transformer = MinMaxScaler()\n",
    "#\n",
    "# Second, the standardized columns\n",
    "stand_columns = ['gene4', 'gene5', 'gene6']\n",
    "# But, we need a two step process: Impute, then standardize\n",
    "steps=[\n",
    "    (\"imputer\", SimpleImputer(missing_values=np.nan, strategy='mean')),\n",
    "    (\"scaler\", StandardScaler())]\n",
    "# Use the pipeline to perform the steps in order\n",
    "stand_transformer = Pipeline(steps)\n",
    "#\n",
    "# Third, the categories\n",
    "categorical_columns = ['gender']\n",
    "categorical_transformer = OrdinalEncoder(categories = [['male','female']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ColumnTransformer(n_jobs=None, remainder='drop', sparse_threshold=0.3,\n",
       "                  transformer_weights=None,\n",
       "                  transformers=[('norm',\n",
       "                                 MinMaxScaler(copy=True, feature_range=(0, 1)),\n",
       "                                 ['gene1', 'gene2', 'gene3']),\n",
       "                                ('stand',\n",
       "                                 Pipeline(memory=None,\n",
       "                                          steps=[('imputer',\n",
       "                                                  SimpleImputer(add_indicator=False,\n",
       "                                                                copy=True,\n",
       "                                                                fill_value=None,\n",
       "                                                                missing_values=nan,\n",
       "                                                                strategy='mean',\n",
       "                                                                verbose=0)),\n",
       "                                                 ('scaler',\n",
       "                                                  StandardScaler(copy=True,\n",
       "                                                                 with_mean=True,\n",
       "                                                                 with_std=True))],\n",
       "                                          verbose=False),\n",
       "                                 ['gene4', 'gene5', 'gene6']),\n",
       "                                ('cat',\n",
       "                                 OrdinalEncoder(categories=[['male', 'female']],\n",
       "                                                dtype=<class 'numpy.float64'>),\n",
       "                                 ['gender'])],\n",
       "                  verbose=False)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now, use the Column Transformer to perform the 3 different groups of transformations\n",
    "preprocessor = ColumnTransformer(\n",
    "    transformers=[\n",
    "        (\"norm\",norm_transformer,norm_columns), # This is a single step transfomer: only norm scale\n",
    "        (\"stand\", stand_transformer, stand_columns), # This is a 2-step transfomer: impute, then stand scale\n",
    "        (\"cat\", categorical_transformer, categorical_columns), # This is a single step, just convert categories\n",
    "    ]\n",
    ")\n",
    "# Show the details of the object\n",
    "preprocessor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.137</td>\n",
       "      <td>0.890</td>\n",
       "      <td>0.766</td>\n",
       "      <td>0.937</td>\n",
       "      <td>1.417</td>\n",
       "      <td>-3.122</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.538</td>\n",
       "      <td>0.293</td>\n",
       "      <td>0.793</td>\n",
       "      <td>-0.307</td>\n",
       "      <td>-1.034</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.336</td>\n",
       "      <td>0.462</td>\n",
       "      <td>0.840</td>\n",
       "      <td>-1.877</td>\n",
       "      <td>-0.210</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.699</td>\n",
       "      <td>0.519</td>\n",
       "      <td>0.315</td>\n",
       "      <td>-0.598</td>\n",
       "      <td>-0.061</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.200</td>\n",
       "      <td>0.841</td>\n",
       "      <td>0.530</td>\n",
       "      <td>0.512</td>\n",
       "      <td>1.050</td>\n",
       "      <td>-2.151</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene1  gene2  gene3  gene4  gene5  gene6  gender\n",
       "0  0.137  0.890  0.766  0.937  1.417 -3.122   0.000\n",
       "1  0.538  0.293  0.793 -0.307 -1.034  0.000   0.000\n",
       "2  0.336  0.462  0.840 -1.877 -0.210  0.000   1.000\n",
       "3  0.699  0.519  0.315 -0.598 -0.061  0.000   0.000\n",
       "4  0.200  0.841  0.530  0.512  1.050 -2.151   1.000"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fit the column transformer\n",
    "preprocessor.fit(X)\n",
    "# Tranform the entire dataframe\n",
    "transformed_array = preprocessor.transform(X)\n",
    "# Build the dataframe, the column order is preserved because we setup the ColumnTransformer from left to right\n",
    "X_transformed = pd.DataFrame(data = transformed_array,columns = X.columns)\n",
    "X_transformed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.591</td>\n",
       "      <td>0.536</td>\n",
       "      <td>0.499</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.208</td>\n",
       "      <td>0.220</td>\n",
       "      <td>0.240</td>\n",
       "      <td>1.005</td>\n",
       "      <td>1.005</td>\n",
       "      <td>1.005</td>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-3.596</td>\n",
       "      <td>-2.329</td>\n",
       "      <td>-3.122</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.472</td>\n",
       "      <td>0.378</td>\n",
       "      <td>0.338</td>\n",
       "      <td>-0.505</td>\n",
       "      <td>-0.756</td>\n",
       "      <td>-0.362</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.654</td>\n",
       "      <td>0.522</td>\n",
       "      <td>0.489</td>\n",
       "      <td>0.093</td>\n",
       "      <td>-0.010</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.733</td>\n",
       "      <td>0.692</td>\n",
       "      <td>0.675</td>\n",
       "      <td>0.501</td>\n",
       "      <td>0.718</td>\n",
       "      <td>0.728</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2.742</td>\n",
       "      <td>2.465</td>\n",
       "      <td>2.277</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene1   gene2   gene3   gene4   gene5   gene6  gender\n",
       "count 100.000 100.000 100.000 100.000 100.000 100.000 100.000\n",
       "mean    0.591   0.536   0.499   0.000  -0.000   0.000   0.550\n",
       "std     0.208   0.220   0.240   1.005   1.005   1.005   0.500\n",
       "min     0.000   0.000   0.000  -3.596  -2.329  -3.122   0.000\n",
       "25%     0.472   0.378   0.338  -0.505  -0.756  -0.362   0.000\n",
       "50%     0.654   0.522   0.489   0.093  -0.010   0.000   1.000\n",
       "75%     0.733   0.692   0.675   0.501   0.718   0.728   1.000\n",
       "max     1.000   1.000   1.000   2.742   2.465   2.277   1.000"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We have successfully prepared the data exactly how we wanted:\n",
    "# genes 1,2,3: standardized\n",
    "# genes 4,5,6: imputed missing values, then standardized\n",
    "# gender: converted to a number\n",
    "X_transformed.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Combine ColumnTransformer with the use of the Pipeline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, let's combine all of this with the pipeline to build a machine learning model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X_train: (80, 7)\n",
      "y_train: (80,)\n",
      "X_test: (20, 7)\n",
      "y_test: (20,)\n"
     ]
    }
   ],
   "source": [
    "# Features\n",
    "X = df.drop(['cancer_detected'],axis = 1)\n",
    "# Target\n",
    "y = df['cancer_detected']\n",
    "# Split into train/test\n",
    "# Reserve 20% for testing\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20,random_state = 42)\n",
    "# Verify the sizes of the split datasets\n",
    "print('X_train:', X_train.shape)\n",
    "print('y_train:', y_train.shape)\n",
    "print('X_test:', X_test.shape)\n",
    "print('y_test:', y_test.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Use the preprocessor we setup above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ColumnTransformer(n_jobs=None, remainder='drop', sparse_threshold=0.3,\n",
       "                  transformer_weights=None,\n",
       "                  transformers=[('norm',\n",
       "                                 MinMaxScaler(copy=True, feature_range=(0, 1)),\n",
       "                                 ['gene1', 'gene2', 'gene3']),\n",
       "                                ('stand',\n",
       "                                 Pipeline(memory=None,\n",
       "                                          steps=[('imputer',\n",
       "                                                  SimpleImputer(add_indicator=False,\n",
       "                                                                copy=True,\n",
       "                                                                fill_value=None,\n",
       "                                                                missing_values=nan,\n",
       "                                                                strategy='mean',\n",
       "                                                                verbose=0)),\n",
       "                                                 ('scaler',\n",
       "                                                  StandardScaler(copy=True,\n",
       "                                                                 with_mean=True,\n",
       "                                                                 with_std=True))],\n",
       "                                          verbose=False),\n",
       "                                 ['gene4', 'gene5', 'gene6']),\n",
       "                                ('cat',\n",
       "                                 OrdinalEncoder(categories=[['male', 'female']],\n",
       "                                                dtype=<class 'numpy.float64'>),\n",
       "                                 ['gender'])],\n",
       "                  verbose=False)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preprocessor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a new Logisitic Regression classification model and use it in a pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logistic Regression Accuracy: 0.9\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[16,  0],\n",
       "       [ 2,  2]])"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the pipeline with our preprocess and a new classifier model\n",
    "#\n",
    "lr = LogisticRegression() # Create a new model\n",
    "# \n",
    "# Use the preprocessor with the model\n",
    "#\n",
    "pipe = Pipeline(\n",
    "    steps=[(\"preprocessor\", preprocessor), (\"LogisticRegressor\", lr)]\n",
    ")\n",
    "\n",
    "# Perform the preprocessing and the training of the model\n",
    "pipe.fit(X_train, y_train)\n",
    "\n",
    "# Treat the pipe object just like trained model\n",
    "y_pred = pipe.predict(X_test)\n",
    "# Report the performance\n",
    "print('Logistic Regression Accuracy:', pipe.score(X_test, y_test))\n",
    "confusion_matrix(y_test, y_pred)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Raw data in, trained and evaluated model out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GBC Accuracy: 0.95\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[16,  0],\n",
       "       [ 1,  3]])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Want to do it again with a different algorithm?\n",
    "#\n",
    "gbc = GradientBoostingClassifier()\n",
    "#\n",
    "# Use the model with the preprocessor\n",
    "#\n",
    "pipe = Pipeline(\n",
    "    steps=[(\"preprocessor\", preprocessor), (\"GBC_classifier\", gbc)]\n",
    ")\n",
    "#\n",
    "# Perform the preprocessing and the training of the model\n",
    "pipe.fit(X_train, y_train)\n",
    "#\n",
    "# Treat the pipe object just like trained model\n",
    "y_pred = pipe.predict(X_test)\n",
    "# Report the performance\n",
    "print('GBC Accuracy:', pipe.score(X_test, y_test))\n",
    "confusion_matrix(y_test, y_pred)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What we did:\n",
    "\n",
    "1. Load the data\n",
    "2. Used ColumnTransfomer to scale/transform one or more columns\n",
    "3. Use a ColumnTransformer to deal with (impute) missing values\n",
    "4. Use a ColumnTranformer to convert categories ('male'/'female') to numbers (0,1)\n",
    "5. Putting several transfomers together to do all transformation with a single ColumnTransfomer\n",
    "6. Combine ColumnTransformer with the use of the Pipeline to transform raw data and fit/evaluate a model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "instance_type": "ml.t3.medium",
  "kernelspec": {
   "display_name": "Python 3 (Data Science)",
   "language": "python",
   "name": "python3__SAGEMAKER_INTERNAL__arn:aws:sagemaker:us-west-2:236514542706:image/datascience-1.0"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
